modules/indexing/src/main/java/org/apache/ignite/internal/processors/query/h2/H2ManagedLocalResult.java

```diff
@@ -43,71 +86,575 @@
      */
     public H2ManagedLocalResult(Session ses, H2MemoryTracker memTracker, Expression[] expressions,
         int visibleColCnt) {
-        super(ses, expressions, visibleColCnt);
+        this.session = ses;
+        rows = Utils.newSmallArrayList();
+        this.visibleColumnCount = visibleColCnt;
+        rowId = -1;
+        this.expressions = expressions;
+        this.memTracker = memTracker;
+
+        if (ses != null && ses.getQueryContext() != null) {
+            QueryContext qctx = (QueryContext)ses.getQueryContext();
 
-        this.mem = memTracker;
+            ctx = qctx.context();
+
+            assert memTracker == null || ctx != null; // Context should be set when we track memory.
+        }
     }
 
-    /** {@inheritDoc} */
-    @Override protected void onUpdate(ValueRow distinctRowKey, Value[] oldRow, Value[] row) {
+    /**
+     * Checks available memory.
+     *
+     * @param distinctRowKey Row key.
+     * @param oldRow Old row.
+     * @param row New row.
+     * @return {@code True} if we have available memory.
+     */
+    private boolean hasAvailableMemory(ValueRow distinctRowKey, Value[] oldRow, Value[] row) {
         assert !isClosed();
-        assert row != null;
 
-        long memory;
+        if (memTracker == null)
+            return true; // No memory management set.
+
+        long memory = calculateMemoryDelta(distinctRowKey, oldRow, row);
+
+        memReserved += memory;
+
+        if (memory < 0) {
+            memTracker.released(-memory);
+
+            return true;
+        }
+        else {
+            try {
+                return memTracker.reserved(memory);
+            }
+            catch (Throwable e) {
+                memReserved -= memory;
```

Comment:
Yes, the insufficient memory budget can be the reason of this exception. I've refactored this snippet to avoid using subtraction in the 'catch' clause.